include/envoy/network/io_handle.h

```diff
@@ -77,6 +77,28 @@ class IoHandle {
    */
   virtual Api::IoCallUint64Result sendmsg(const Buffer::RawSlice* slices, uint64_t num_slice,
                                           int flags, const Address::Instance& address) PURE;
+
+  /**
+   * Receive a message into given slices, output overflow, source/destination
+   * addresses via passed-in parameters upon success.
+   * @param slices points to the location of receiving buffer.
+   * @param num_slice indicates number of slices |slices| contains.
+   * @param self_port the port this handle is assigned to. This is used to populate
+   * local_address because local port can't be retrieved from control message.
+   * @param dropped_packets number of packets dropped by kernel because of
+   * receiving buffer overflow. Modified if not nullptr and there is overflow between
+   * this call and previous one.
+   * @param local_address the destination address from transport header.
+   * Modified upon each call.
+   * @param peer_address the source address from transport header. Modified upon
+   * each call.
+   * @return a Api::IoCallUint64Result with err_ = an Api::IoError instance or
+   * err_ = nullptr and rc_ = the bytes received for success.
+   */
+  virtual Api::IoCallUint64Result
+  recvmsg(Buffer::RawSlice* slices, const uint64_t num_slice, uint32_t self_port,
```

Comment:
IMO It's hard to reason about all of the output parameters. Would it be better to return a struct which uses absl::optional<> where needed? I could go either way but just throwing this out there for consideration.